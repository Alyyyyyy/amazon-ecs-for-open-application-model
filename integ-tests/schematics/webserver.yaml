apiVersion: core.oam.dev/v1alpha1
kind: ApplicationConfiguration
metadata:
  name: webserver-app
  annotations:
    version: v1.0.0
    description: "Example webserver app"
spec:
  variables:
    - name: message
      value: "Well hello there"
  components:
    - componentName: frontend
      instanceName: web-front-end
      parameterValues:
        - name: message
          value: "[fromVariable(message)]"

    - componentName: backend
      instanceName: backend-svc
---
apiVersion: core.oam.dev/v1alpha1
kind: ComponentSchematic
metadata:
  name: frontend
  annotations:
    version: v1.0.0
    description: "A simple webserver"
spec:
  workloadType: core.oam.dev/v1alpha1.Server
  parameters:
    - name: message
      description: The message to display in the web app.
      type: string
      value: "Hello from my app, too"
  containers:
    - name: web
      env:
        - name: MESSAGE
          fromParam: message
      image: example/frontend-svc:latest
      imagePullSecret: dockerhub-creds
      resources:
        cpu:
          required: 1.5
        memory:
          required: 128M
      ports:
        - name: port
          containerPort: 80
---
apiVersion: core.oam.dev/v1alpha1
kind: ComponentSchematic
metadata:
  name: backend
  annotations:
    version: v1.0.0
    description: "A backend webserver"
spec:
  workloadType: core.oam.dev/v1alpha1.Server
  containers:
    - name: web
      image: example/backend-api:latest
      imagePullSecret: dockerhub-creds
      resources:
        cpu:
          required: 1.25
        memory:
          required: 768M
      ports:
        - name: port
          containerPort: 4000
          protocol: UDP
    - name: sidecar
      image: example/backend-sidecar:latest
      imagePullSecret: other-dockerhub-creds
      resources:
        cpu:
          required: 0.5
        memory:
          required: 128M
      ports:
        - name: port
          containerPort: 4001
          protocol: TCP
