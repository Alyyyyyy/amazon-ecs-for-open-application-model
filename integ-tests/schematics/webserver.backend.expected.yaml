AWSTemplateFormatVersion: 2010-09-09
Description: Amazon ECS infrastructure for webserver-app backend-svc

Resources:
  LogGroup:
    Type: AWS::Logs::LogGroup
    Properties:
      LogGroupName: oam-ecs-webserver-app-backend-svc

  TaskDefinition:
    Type: AWS::ECS::TaskDefinition
    Properties:
      Family: oam-ecs-webserver-app-backend-svc
      NetworkMode: awsvpc
      RequiresCompatibilities:
        - FARGATE
      Cpu: 2.00 vcpu
      Memory: '4096'
      ExecutionRoleArn: !GetAtt ExecutionRole.Arn
      ContainerDefinitions:
        - Name: web
          Image: example/backend-api:latest
          PortMappings:
            - ContainerPort: 4000
              Protocol:  udp
          RepositoryCredentials:
            CredentialsParameter: "dockerhub-creds"
          LogConfiguration:
            LogDriver: awslogs
            Options:
              awslogs-region: !Ref AWS::Region
              awslogs-group: !Ref LogGroup
              awslogs-stream-prefix: oam-ecs
        - Name: sidecar
          Image: example/backend-sidecar:latest
          PortMappings:
            - ContainerPort: 4001
              Protocol:  tcp
          RepositoryCredentials:
            CredentialsParameter: "other-dockerhub-creds"
          LogConfiguration:
            LogDriver: awslogs
            Options:
              awslogs-region: !Ref AWS::Region
              awslogs-group: !Ref LogGroup
              awslogs-stream-prefix: oam-ecs

  ExecutionRole:
    Type: AWS::IAM::Role
    Properties:
      AssumeRolePolicyDocument:
        Statement:
          - Effect: Allow
            Principal:
              Service: ecs-tasks.amazonaws.com
            Action: 'sts:AssumeRole'

      Policies:
        - PolicyName: PrivateRegistryCreds
          PolicyDocument:
            Version: '2012-10-17'
            Statement:
              - Effect: 'Allow'
                Action:
                  - 'secretsmanager:GetSecretValue'
                Resource:
                  - !Sub 'arn:${AWS::Partition}:secretsmanager:${AWS::Region}:${AWS::AccountId}:secret:dockerhub-creds-??????'
                  - !Sub 'arn:${AWS::Partition}:secretsmanager:${AWS::Region}:${AWS::AccountId}:secret:other-dockerhub-creds-??????'
      ManagedPolicyArns:
        - !Sub 'arn:${AWS::Partition}:iam::aws:policy/service-role/AmazonECSTaskExecutionRolePolicy'

  ContainerSecurityGroup:
    Type: AWS::EC2::SecurityGroup
    Properties:
      GroupDescription: oam-ecs-webserver-app-backend-svc-ContainerSecurityGroup
      VpcId:
        Fn::ImportValue: oam-ecs-VpcId

  Service:
    Type: AWS::ECS::Service
    Properties:
      Cluster:
        Fn::ImportValue: oam-ecs-ECSCluster
      TaskDefinition: !Ref TaskDefinition
      DeploymentConfiguration:
        MinimumHealthyPercent: 100
        MaximumPercent: 200
      DesiredCount: 1
      LaunchType: FARGATE
      NetworkConfiguration:
        AwsvpcConfiguration:
          AssignPublicIp: DISABLED
          Subnets:
            Fn::Split:
              - ','
              - Fn::ImportValue: oam-ecs-PrivateSubnets
          SecurityGroups:
            - !Ref ContainerSecurityGroup
      LoadBalancers:
        - ContainerName: web
          ContainerPort: 4000
          TargetGroupArn: !Ref TargetGroupWeb4000
        - ContainerName: sidecar
          ContainerPort: 4001
          TargetGroupArn: !Ref TargetGroupSidecar4001
      HealthCheckGracePeriodSeconds: 0
    DependsOn:
      - LBListenerWeb4000

      - LBListenerSidecar4001



  SGLoadBalancerToContainers:
    Type: AWS::EC2::SecurityGroupIngress
    Properties:
      Description: Ingress from anywhere on the internet through the public NLB
      GroupId: !Ref ContainerSecurityGroup
      IpProtocol: '-1'
      CidrIp: 0.0.0.0/0

  PublicLoadBalancer:
    Type: AWS::ElasticLoadBalancingV2::LoadBalancer
    Properties:
      Type: network
      Scheme: internet-facing
      Subnets:
        Fn::Split:
          - ','
          - Fn::ImportValue: oam-ecs-PublicSubnets

  LBListenerWeb4000:
    Type: AWS::ElasticLoadBalancingV2::Listener
    Properties:
      DefaultActions:
        - TargetGroupArn: !Ref TargetGroupWeb4000
          Type: 'forward'
      LoadBalancerArn: !Ref 'PublicLoadBalancer'
      Port: 4000
      Protocol:  UDP

  TargetGroupWeb4000:
    Type: AWS::ElasticLoadBalancingV2::TargetGroup
    Properties:
      Protocol:  UDP
      TargetType: ip
      Port: 4000
      VpcId:
        Fn::ImportValue: oam-ecs-VpcId
      TargetGroupAttributes:
      - Key: deregistration_delay.timeout_seconds
        Value: '30'


  LBListenerSidecar4001:
    Type: AWS::ElasticLoadBalancingV2::Listener
    Properties:
      DefaultActions:
        - TargetGroupArn: !Ref TargetGroupSidecar4001
          Type: 'forward'
      LoadBalancerArn: !Ref 'PublicLoadBalancer'
      Port: 4001
      Protocol:  TCP

  TargetGroupSidecar4001:
    Type: AWS::ElasticLoadBalancingV2::TargetGroup
    Properties:
      Protocol:  TCP
      TargetType: ip
      Port: 4001
      VpcId:
        Fn::ImportValue: oam-ecs-VpcId
      TargetGroupAttributes:
      - Key: deregistration_delay.timeout_seconds
        Value: '30'



Outputs:
  CloudFormationStackConsole:
    Description: The AWS console deep-link for the CloudFormation stack
    Value: !Sub https://console.aws.amazon.com/cloudformation/home?region=${AWS::Region}#/stacks/stackinfo?stackId=${AWS::StackName}

  ECSServiceConsole:
    Description: The AWS console deep-link for the ECS service
    Value: !Sub https://console.aws.amazon.com/ecs/home?region=${AWS::Region}#/clusters/oam-ecs/services/${Service.Name}

  WebPort4000Endpoint:
    Description: The endpoint for container Web on port 4000
    Value: !Sub '${PublicLoadBalancer.DNSName}:4000'

  SidecarPort4001Endpoint:
    Description: The endpoint for container Sidecar on port 4001
    Value: !Sub '${PublicLoadBalancer.DNSName}:4001'

